### ISSUE TITLE
Tutorial 6: How to Improve Model's Performance?

### ISSUE THREAD

**Issue by Ninja16180** [bentrevett/pytorch-seq2seq#108]
Hi Ben,

First of all apology for bugging you multiple times with my queris/issues faced during training the transformer model with my dataset following tutorial 6.
Your help has been vary valuable to complete the training of the model without any error.

However, it is the performance of the trained model, which made me to seek your guidance once again.

My objective is to make a converational model and the encoder and decoder data have set of sentences and corresponding responses.

Size of dataset:
The dataset has 71K rows of sentence-response pairs.
Out of which 10% has been used as validation dataset and another 10% as test dataset.

I ran the training for more than 10 epochs (13 epochs in total).
But unfortunately the loss remains quite high at the end of the last epoch:

```
Epoch: 13 | Time: 6m 22s
	Train Loss: 2.588 | Train PPL:  13.300
	 Val. Loss: 6.178 |  Val. PPL: 482.263
```

Also, while tested with random sentences (from training dataset itself) to check the response, the model has been giving away one common sentence as the response for all input sentences:
`predicted response = ['i', 'am', 'not', 'sure', 'what', 'you', 'are', 'doing', '<eos>']`

I am not quite sure where I need to tweak to improve the model's performance so that the training loss will be less and the model could ouput proper response as expected.

I am kind of a novish in this field and so far I have only been learning by following various tutorials as it is. 
This is for the first time I am trying with my customised data and thus came across some challenges which you have helped in resolving.
It is a lot to ask for as you have already been selflessly helping the community to learn as well as to resolve the issues they face along the way; still I would really appreciate your mentorship to help me to achieve this last bit of my attempt of building a conversational model.

Here is the github repo where I uploaded the jupyter file with end to end result:
https://github.com/Ninja16180/Rough_Transformer/blob/master/Rough_Transformer_v.05.ipynb

**Comment by bentrevett** [comment]
Unfortunately the hyperparameters used in the tutorial notebook are tuned especially for that dataset and there is no guarantee they are the optimal parameters for your dataset. This means you'll have to do a bit of hyperparameter tuning - so all of: learning rate, hidden dimension, layers, heads, pff dim. and dropout will have to be tuned.

I'd suggest making a small script that chooses a random value for each from a list of sensible values, i.e. `hid_dim = random.choice([64, 128, 256, 512])`, and having this script repeatedly choose values, run the model for 10 or so epochs and write the results somewhere where you can check all the values and see which setting gives you the best results. This is time consuming, but there isn't really a quick way of doing it.

Having the model output the same sentence for every single input is very odd but usually means that your model is ignoring the actual input sentence and has just learned to output a very safe sentence. I believe a better hyperparameter choice should solve this.

Some suggested values are:
```
learning rate = [1e-3, 5e-4, 1e-4, 5e-5, 1e-5] #1e-3 might be too high
layers = [2,3,4,5]
heads = [5, 10, 15]
pf. dim. = [hid dim, hid dim * 2, hid dim * 3] #this should usually be larger than hid dim
dropout = [0, 0.1, 0.2, 0.3, 0.4, 0.5]
```
Maybe start with these choices and if one option always leads to bad results, say a learning rate of 1e-3, then remove it from the list.

One other thing is that the original Transformer model was actually trained with a different optimizer that has a gradual "warm-up" phase where it increases from a very small value, followed by a "cool down" phase where it exponentially decays, see: https://nlp.seas.harvard.edu/2018/04/03/attention.html#optimizer.

If the hyperparameter search doesn't work, I would try switching to the optimizer in the article above or something similar, such as the `OneCycleLR` scheduler provided by PyTorch: https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.OneCycleLR.

**Comment by Ninja16180** [comment]
Hi Ben,

Thanks a lot for such a detailed guidance; I will start working towards tuning the hyperparameters for my dataset.

Looks like quite a job to be done in here; but it's definitely going to be a great learning for a newbie like me.

I know, now it's going out of scope from this discussion forum as now it is more of dealing with my own dataset only, still I would like to request you for your valuable guidance along the way. So far it has been only self learning and this is for the first time I have been receiving expert's help exactly when it is very much required to move to the next step. 

I will try tuning the hyperparameters by myself, but can I please reach out to you if I require guidance where I am completely stuck?

I can understand it might not be the right platform to ask for this kind of help and already you have been helping so much completely selflessly.
Please let me know if there is any other forum where you take up this type of guidance requests and I will drop my query there. 
Being a newbie and self learner in this field, I sometimes find it bit challenging to move forward and for the first time I was fortunate to get mentorship from an expert like you who helped me in this situation.

Thanks a lot for all your help.
